### ISSUE TITLE
Segfault on PangoCairo.create_layout

### ISSUE THREAD

**Issue by rakeshvar** [Kozea/cairocffi#48]
``` python
python3 -c "
> from gi.repository import Gtk, Pango, PangoCairo
> import cairocffi as cairo
> surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, 120, 120)
> context = cairo.Context(surface)
> layout = PangoCairo.create_layout(context)
> "
Segmentation fault (core dumped)
```

I do not get the same segfault when using cairo from pycairo.

**Comment by rakeshvar** [comment]
I tried the entire procedure of installing pygobject and cairocffi on a different machine, and still got the same seg fault! Could anybody else reproduce this?

**Comment by SimonSapin** [comment]
Sorry, I thought I had answered this the first time…

TL;DR: use [this](http://pythonhosted.org/cairocffi/cffi_api.html#converting-cairocffi-wrappers-to-pycairo).

What’s going on is that `PangoCairo.create_layout` expects to be given a `Context` object from pycairo. It assumes that the given object’s layout in memory is [pycairo’s `PycairoContext` C struct](http://cgit.freedesktop.org/py2cairo/tree/src/pycairo.h?id=f3435910d8f5365b45ebd4216f4043383c9e3e19#n40) and tries to access the `cairo_t *ctx` field to get a pointer. cairocffi’s `Context` object’s internal representation is completely different, so PyGObjects gets random data, assumes it’s a pointer, and eventually segfaults probably when trying to dereference it.

There is [some discussion](https://bugzilla.gnome.org/show_bug.cgi?id=707196) about making PyGObject support cairocffi, but not much has happened so far.

In the meantime, you can use [unsafe pointer manipulation tricks](http://pythonhosted.org/cairocffi/cffi_api.html#converting-cairocffi-wrappers-to-pycairo) to get a pycairo `Context` for the same `cairo_t` cairo context as a given cairocffi `Context`. This kinda defeats the point of using cairocffi in the first place since you still have to also use pycairo, but that’s all I have for now.

**Comment by liZe** [comment]
That's not really a bug in CairoCFFI, closing as it's documented.

**Comment by artscoop** [comment]
The issue here is not documented anymore (404 for the links posted by @SimonSapin), so I'm not sure what are the resolution steps for this.

**Comment by SimonSapin** [comment]
New docs URL: http://cairocffi.readthedocs.io/en/latest/cffi_api.html#converting-cairocffi-wrappers-to-pycairo
